src/aliceVision/image/caching.hpp

```diff
@@ -0,0 +1,435 @@
+ * or until there is nothing to remove
+ * 5. if the image fits in the maximal size, load it, store it and return it
+ * 6. the image is too big for the cache, throw an error.
+ * 
+ * In the process described above, we also take advantage of the cache content when loading an image: 
+ * if the same image with a lower half-sampling level (i.e. higher resolution) exists in the cache, 
+ * we take the high-resolution version of the image from the cache and create and new downscaled version of it
+ * instead of loading the image from disk.
+ */
+class ImageCache 
+{
+public:
+    /**
+     * @brief Create a new image cache by defining memory usage limits and image reading options.
+     * @param[in] capacity_MB the cache capacity (in MB)
+     * @param[in] maxSize_MB the cache maximal size (in MB)
+     * @param[in] options the reading options that will be used when loading images through this cache
+     */
+    ImageCache(float capacity_MB, float maxSize_MB, const ImageReadOptions& options);
+
+    /**
+     * @brief Destroy the cache and the unused images it contains.
+     */
+    ~ImageCache();
+
+    /// make image cache class non-copyable
+    ImageCache(const ImageCache&) = delete;
+    ImageCache& operator=(const ImageCache&) = delete;
+
+    /**
+     * @brief Retrieve a cached image at a given half-sampling level.
+     * @note This method is thread-safe.
+     * @param[in] filename the image's filename on disk
+     * @param[in] halfSampleLevel the half-sampling level
+     * @return a shared pointer to the cached image
+     * @throws std::runtime_error if the image does not fit in the maximal size of the cache
+     */
+    template<typename TPix>
+    std::shared_ptr<Image<TPix>> get(const std::string& filename, int halfSampleLevel);
+
+    /**
+     * @return information on the current cache state and usage
+     */
+    inline const CacheInfo& info() const
+    {
+        return _info;
+    }
+
+    /**
+     * @return the image reading options of the cache
+     */
+    inline const ImageReadOptions& readOptions() const
+    {
+        return _options;
+    }
+
+    /**
+     * @brief Provide a description of the current internal state of the cache (useful for logging).
+     * @return a string describing the cache content
+     */
+    std::string toString() const;
+
+private:
+    /**
+     * @brief Load a new image corresponding to the given key and add it as a new entry in the cache.
+     * @param[in] key the key used to identify the entry in the cache
+     */
+    template<typename TPix>
+    void load(const CacheKey& key);
+
+    CacheInfo _info;
+    ImageReadOptions _options;
+    std::unordered_map<CacheKey, CacheValue, CacheKeyHasher> _imagePtrs;
+    /// ordered from LRU (Least Recently Used) to MRU (Most Recently Used)
+    std::list<CacheKey> _keys;
+    mutable std::mutex _mutex;
+
+};
+
+
+// Since some methods in the ImageCache class are templated
```

Comment:
Maybe we can do like `image\io.hpp` and use a templated function in cpp to avoid huge hpp.